### PR TITLE
[DO NOT MERGE] Refactor & Upgrade VMC/Gateway logic

### PR DESCRIPTION
**Pull request by gakonst**
- [x] Create CrossChain class which is to be used as a base class for any crosschain operations, such as Gateways or Plasma. It gets initialized with a dappchain and ethereum key.
- [x] Create GatewayUser which extends CrossChain for gateways
- [x] Separate DPOSUser from CrossChain
- [x] Update gateway bindings for https://github.com/loomnetwork/transfer-gateway-v2/pull/83
- [x] Add typescript bindings for ethers contracts
- [x] Update protobufs once we switch how the WithdrawalReceipt is stored on LoomChain (https://github.com/loomnetwork/go-loom/pull/271)
- [x] Reset and cleanup commits

(DO NOT MERGE - THIS IS NOT BACKWARDS COMPATIBLE, ONLY MERGE AFTER DPOS-UI HAS BEEN UPDATED)

### CODE REVIEW

File: src/gateway-user.ts [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+  static async createGatewayOfflineUserAsync(
+    endpoint: string,
+    privateKey: string,
+    dappchainEndpoint: string,
+    dappchainKey: string,
+    chainId: string,
+    gatewayAddress: string,
+    vmcAddress: string,
+    loomAddress: string
+  ): Promise<GatewayUser> {
```
Comment (enlight) [inline src/gateway-user.ts]: This is a ridiculous number of positional arguments, add an interface type
```typescript
interface ICreateGatewayOfflineUserArgs {
    endpoint: string
    privateKey: string
    dappchainEndpoint: string
    dappchainKey: string
    chainId: string
    gatewayAddress: string
    vmcAddress: string
    loomAddress: string
}

static async createGatewayOfflineUserAsync({
    endpoint,
    privateKey,
    dappchainEndpoint,
    dappchainKey,
    chainId,
    gatewayAddress,
    vmcAddress,
    loomAddress,
}: ICreateGatewayOfflineUserArgs): Promise<GatewayUser> {
```

Comment (enlight) [reply src/gateway-user.ts]: Oh, and this doesn't need `async`

File: package.json [hunk 2]
```diff
@@ -49,6 +49,7 @@
     "shelljs": "^0.8.2",
     "tslib": "^1.9.3",
     "tweetnacl": "^1.0.0",
+    "typechain": "^0.3.11",
```
[the remark this answers is not in the document]
Comment (gakonst) [reply package.json]: Typed bindings for ethereum contracts, makes life easier https://blog.neufund.org/introducing-typechain-typescript-bindings-for-ethereum-smart-contracts-839fc2becf22

**Comment by gakonst** [comment]
I'm going to split this PR into two and let the refactor for later.

**Pull request closed without merging**